File: 1776-minimum-operations-to-reduce-x-to-zero/minimum-operations-to-reduce-x-to-zero.cpp
class Solution {
public:
    int minOperations(vector<int>& nums, int x) {
        int n=nums.size();
        unordered_map<int,int>pre;
        unordered_map<int,int>suff;
        int p=0;
        int s=0;
        for(int i=0;i<n;i++){
            p+=nums[i];
            pre[p]=i;
            s+=nums[n-1-i];
            suff[s]=i;
        }
        int np=INT_MAX;
        int cp=INT_MAX;
        int ans=INT_MAX;
        if(pre.find(x)!=pre.end()){
            np=pre[x]+1;
        }
        if(suff.find(x)!=suff.end()){
            cp=suff[x]+1;
        }
        ans=min(np,cp);
        
        for(auto it:pre){
            
           if(suff.find(x-it.first)!=suff.end()){
         
                ans=min(ans,it.second+suff[x-it.first]+2); 
                
           } 
        }
        if(ans==INT_MAX){
            return -1;
        }
        if(ans>n){
           return -1;
        }
        return ans;
    }
};

File: 1776-minimum-operations-to-reduce-x-to-zero/minimum-operations-to-reduce-x-to-zero.java
class Solution {
    public int minOperations(int[] nums, int x) {
       int n=nums.length;
       // vector<int>pre;
       ArrayList<Integer>pre=new ArrayList<>();
        HashMap<Integer,Integer>suff=new HashMap<>();
        int p=0, s=0;
        int ans =Integer.MAX_VALUE;
        for(int i=0;i<n;i++){
            p+=nums[i];
            pre.add(p);
            s+=nums[n-1-i];
            suff.put(s,i);
            if(p==x || s==x){
                ans =Math.min(ans,i+1);
            }
        }

        for(int i=0; i<n; i++){
            int req = x-pre.get(i);
            if(suff.containsKey(req)){
                int op = i+1+suff.get(req)+1;
                ans = Math.min(ans,op);
            }
        }

        if(ans ==Integer.MAX_VALUE){
            return -1;
        }
        if(ans>n){
            return -1;
        }
        return ans;  
    }
}